feat(Submit): Improve and add action menu

coder3101 opened the pull request:
This commit improves and add following:
1. "Submit" in Action menu
2. Shortcut for "Submit", default is <kbd>Ctrl+Alt+S</kbd>
3. Improve Submit logic, and warns when submission through old cf tool.

## Related Issues / Pull Requests
Fixes #679 and Fixes #201 

## Motivation and Context
This is minor changes to CF Tool and submission.

## How Has This Been Tested?
Arch Linux with KDE

## Screenshots (if appropriate)

## Checklist
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- You can open a pull request before all these are done, but they should be done before getting merged. -->
- [x] If the key of a setting is changed, the `old` attribute is updated or it is resolved in SettingsUpdater.
- [x] If there are changes of the text displayed in the UI, they are wrapped in `tr()` or `QCoreApplication::translate()`.
- [x] If needed, I have opened a pull request or an issue to update the [documentation](https://github.com/cpeditor/cpeditor.github.io).
- [x] If these changes are notable, they are documented in [CHANGELOG.md](https://github.com/cpeditor/cpeditor/blob/master/CHANGELOG.md).

## Additional text
In future we will re-write submission logic, till then either this PR gets merged or updated to reflect the changes.

coder3101 committed:
feat(Submit/CF Tool): Improve and add action menu

coder3101 committed:
add changelog and update Translation

ouuan commented on src/mainwindow.cpp:
```diff
@@ -701,6 +701,18 @@ void MainWindow::compileAndRun()
+void MainWindow::submitSolution()
+{
+    LOG_INFO("Requesting submit solution");
+    if (submitToCodeforces)
```
```suggestion
    if (canSubmitSolution())
```

Reuse codes when possible.

ouuan commented on ui/appwindow.ui:
```diff
@@ -491,6 +492,14 @@
+    <string>Submit</string>
+   </property>
+   <property name="shortcut">
+    <string>Ctrl+Alt+S</string>
```
```suggestion
    <string notr="true">Ctrl+Alt+S</string>
```

ouuan reviewed (changes requested):
1. I don't like the "create when needed" logic. I'd rather show it when needed and hide it when not available.

2. I remember that CF Tool fails to submit to URLs with problem code "0" which is in the URL when you enter a running contest, which is the reason I parsed the URL for CF Tool 1.0. I think we can pass the URL to CF Tool only when we fail to parse the URL.

coder3101 committed:
Apply suggestions from code review

Co-authored-by: Yufan You <ouuansteve@gmail.com>

coder3101 committed:
Merge branch 'master' into submit-in-action-menu

coder3101 committed:
update translations

coder3101 committed:
Rephrase the CHANGELOG

coder3101 commented on CHANGELOG.md:
```diff
@@ -6,10 +6,13 @@
 
--   Fixed an issue that in some rare cases CP Editor could not get full data from Competitive Companion. (#345 and #624)
+-   Fixed an issue that in some rare cases CP Editor could not get full data from Competitive Companion. (#345 and #624) 
+-  	Fixed an issue where some URL was not being parsed by CF Tool. (#679 and #681) 
```
```suggestion
-   Fixed an issue where some URL was not being parsed by CF Tool. (#679 and #681) 
```

coder3101 committed:
Update CHANGELOG.md

coder3101 commented:
> I remember that CF Tool fails to submit to URLs with problem code "0" which is in the URL when you enter a running contest, which is the reason I parsed the URL for CF Tool 1.0. I think we can pass the URL to CF Tool only when we fail to parse the URL.

Yup, I see it is still open.
https://github.com/xalanq/cf-tool/issues/79

What about this solution instead. 

If url ends with "0" we make it "A" and submit with that new modified url.

coder3101 committed:
Fix problem code 0 for cf tool 1.0 or above

coder3101 committed:
Revert changed problem code warning message.

coder3101 committed:
Update translations

coder3101 committed:
Submit button is set visible or hidden depening on URL

coder3101 committed:
Improve some code

github-actions[bot] committed:
style: Format codes (#684)

Co-authored-by: coder3101 <coder3101@users.noreply.github.com>

coder3101 committed:
Format code in settings.json

ouuan commented on src/Extensions/CFTool.cpp:
```diff
@@ -57,52 +57,78 @@ void CFTool::submit(const QString &filePath, const QString &url)
+    CFToolProcess->setProgram(CFToolPath);
+
+    bool wasProblemCodeChanged = false;
+    QString urlCopy = url;
```
Declare it when it's used, and give it a meaningful name like `convertedUrl`.

coder3101 committed:
Apply code review suggestion

coder3101 committed:
Merge branch 'master' into submit-in-action-menu

coder3101 committed:
Use lambda for connecting signals

IZOBRETATEL777 committed:
Update ru_RU.ts

coder3101 commented:
@ouuan Anythine else you want me to update in this PR?

ouuan committed:
Merge branch 'master' into submit-in-action-menu

ouuan committed:
Update translations and text

coder3101 committed:
Merge branch 'master' into submit-in-action-menu

coder3101 committed:
Merge branch 'master' into submit-in-action-menu

IZOBRETATEL777 committed:
add Russian translations

ouuan reviewed (changes requested):
`MainWindow::cftoolPath`, `CFTool::problemContestId`, and `CFTool::problemCode` are not properly initialized, which resulted in:
1. always show "CF Tool not installed" warning
2. empty contest ID and problem code in the toast message.

My suggestion is to wait for #212. You can keep the action in the menu and hotkey part and revert the CF Tool updates. If you still want to update the CF Tool logic, you need to WELL-TEST it.

coder3101 committed:
Merge branch 'master' into submit-in-action-menu

coder3101 committed:
update translations

coder3101 committed:
CFtool: Improve logic and refactor

1. Refactor memebers so that they are more readable and sensible
2. Show CFTool error only when submitting the solution, eliminate the
   pre-check
3. Deprecate the use of CF tool < 1.0 as its support should be removed in next
   release cycle. (Reason: CF tool was beta and progressively it added
   URL support and now it is stable and to avoid legacy code to support
   it. We deprecate it now, In next release cycle, if we detech version < 1.0, we fail the submission requesting an update. In this way we do not have to worry about URL parsing and leverage this task to the tool that does the submission)
4. Improve code and re-usability.
5. Submit button will be enabled always, If it is not installed, it will
   show error during submission, Submit button and action will be hidden
   from tabs that do not have CF URL.

github-actions[bot] committed:
style: format codes (#758)

Co-authored-by: coder3101 <coder3101@users.noreply.github.com>

coder3101 committed:
Update translations

coder3101 commented:
After this code makes to beta, the next alpha/master will have a commit where we will completely drop the support for CF Tool < 1.0 because of the following:

1. CF Tool 0.x.x were all beta and prone to breaking changes. 1.0 should be considered at least API/cmd stable.
2. To avoid having legacy code to Parse URL which can change any day from CF, isn't actually our issue, any such changes should be delegated to CF Tool.

As a result of above, this commit adds a warning to the user that they need to update cf tool to 1.0 or above, as in future release we will not support it.

coder3101 committed:
Merge branch 'master' into submit-in-action-menu

coder3101 committed:
Update CHANGELOG

coder3101 committed:
Update CHANGELOG.md

ouuan commented:
>     empty contest ID and problem code in the toast message.

Haven't read the code yet. But this problem remains.

coder3101 commented:
Sure! I will fix it.

coder3101 committed:
Merge branch 'master' into submit-in-action-menu

coder3101 committed:
Move cfParseUrl to utils. Fix missing toast problem code/id

coder3101 committed:
update translations

coder3101 committed:
Format code

coder3101 commented:
> > empty contest ID and problem code in the toast message.
> 
> Haven't read the code yet. But this problem remains.

Fixed. Please review now!

coder3101 commented on src/Extensions/CFTool.cpp:
```diff
@@ -27,125 +28,123 @@
+        if (url.endsWith("0"))
+        {
+            wasProblemCodeChanged = true;
+            convertedUrl = url.mid(0, url.size() - 1) + "A";
```
```suggestion
            lastUrl = convertedUrl = url.mid(0, url.size() - 1) + "A";
```

coder3101 commented on src/Extensions/CFTool.cpp:
```diff
@@ -27,125 +28,123 @@
-        CFToolProcess->start();
-        bool started = CFToolProcess->waitForStarted(2000);
-        if (started)
+    // @coder3101: Please delete this branch when CFTool < 1.0 is no longer supported.
```
```suggestion
    // @coder3101: Please delete this branch and wasProblemCodeChanged, when CFTool < 1.0 is unsupported.
```

coder3101 commented on src/appwindow.cpp:
```diff
@@ -886,6 +891,8 @@ void AppWindow::onTabChanged(int index)
 
     reAttachLanguageServer(tmp);
 
+    ui->actionSubmit->setEnabled(tmp->canSubmitSolution());
```
```suggestion
    ui->actionSubmit->setVisible(tmp->canSubmitSolution());
```

coder3101 committed:
Apply suggestions from code review

coder3101 committed:
Merge branch 'master' into submit-in-action-menu

coder3101 committed:
Merge branch 'master' into submit-in-action-menu
